sdks/go/pkg/beam/core/graph/fn.go

```diff
@@ -847,9 +886,15 @@ func validateSdfSigNumbers(fn *Fn, num int) error {
 func validateSdfSigTypes(fn *Fn, num int) error {
 	restrictionT := fn.methods[createInitialRestrictionName].Ret[0].T
 	rTrackerT := reflect.TypeOf((*sdf.RTracker)(nil)).Elem()
-
-	for _, name := range requiredSdfNames {
-		method := fn.methods[name]
+	bRTrackerT := fn.methods[createTrackerName].Ret[0].T
+	optionalSdfs := optionalSdfNameMap()
+
+	for _, name := range sdfNames {
+		method, ok := fn.methods[name]
+		if !ok && optionalSdfs[name] {
+			// skip validating unimplemented optional sdf methodsß
+			continue
+		}
```

Comment:
I don't think we should be doing this kind of filtering here. If we need it, we can use requiredSdfNames instead of sdfNames as suggested above, but more importantly, we actually do want the optional sdf functions to be validated (that's why you've got your truncateRestrictionName case below).

Reply:
Ah, I missed that this would throw if the method is unimplemented before your changes - I understand why you're doing this now.

With that said, I think it might be cleaner to move the optional sdfs into their own switch statement rather than doing this kind of filtering. That also lets us keep it so that this function will throw if a required sdf name isn't present (I think other validation does check that, so its probably not a big deal either way). That would let you get rid of this map entirely and you wouldn't have to do any special filtering.

I'll leave that as an optional suggestion

Reply:
That makes sense. If additonal optional sdf methods are added in the future then it will make the existing switch much larger. So I guess, I'll have a separate switch for optional SDFs.